source/postprocess/visualization/mass_error.cc

```diff
@@ -0,0 +1,175 @@
+                  in.velocity);
+              fe_values_support[this->introspection().extractors.temperature].get_function_values (this->get_solution(),
+                  in.temperature);
+              fe_values_support[this->introspection().extractors.pressure].get_function_values (this->get_solution(),
+                  in.pressure);
+              fe_values_support[this->introspection().extractors.pressure].get_function_gradients (this->get_solution(),
+                  in.pressure_gradient);
+              for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                fe_values_support[this->introspection().extractors.compositional_fields[c]].get_function_values(this->get_solution(),
+                    composition_values[c]);
+
+              in.position = fe_values_support.get_quadrature_points();
+
+              // since we are not reading the viscosity and the viscosity
+              // is the only coefficient that depends on the strain rate,
+              // we need not compute the strain rate. set the corresponding
+              // array to empty, to prevent accidental use and skip the
+              // evaluation of the strain rate in evaluate().
+              in.strain_rate.resize(0);
+
+              for (unsigned int i=0; i<support_points.size(); ++i)
+                {
+                  for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                    in.composition[i][c] = composition_values[c][i];
+                }
+              in.cell = &cell;
+
+              this->get_material_model().evaluate(in, out);
+
+              std::vector<Tensor<1,dim> > velocity(n_q_points);
+              std::vector<double>         velocity_divergence(n_q_points);
+              fe_values[this->introspection().extractors.velocities].get_function_values(this->get_solution(),velocity);
+              fe_values[this->introspection().extractors.velocities].get_function_divergences(this->get_solution(),velocity_divergence);
+
+              double rho_div_u = 0.0;
+              double u_grad_rho = 0.0;
+              double cell_volume = 0.0;
+
+              for (unsigned int q = 0; q < n_q_points; ++q)
+                {
+                  Tensor<1,dim> density_gradient = Tensor<1,dim>();
+                  double        density = 0.0;
+                  for (unsigned int i = 0; i < support_points.size(); ++i)
+                    {
+                      density_gradient += out.densities[i] * fe_values_density.shape_grad(i,q);
+                      density += out.densities[i] * fe_values_density.shape_value(i,q);
+                    }
+
+                  rho_div_u  += density * velocity_divergence[q] * fe_values.JxW(q);
+                  u_grad_rho += velocity[q] * density_gradient * fe_values.JxW(q);
+                  cell_volume += fe_values.JxW(q);
+                }
+
+              (*return_value.second)(cell_index) = (rho_div_u + u_grad_rho) / cell_volume;
```

Comment:
OK, so you compute *average* values on each cell. This should be part of the documentation.